tests/unit/s2n_tls13_cert_verify_test.c

```diff
@@ -236,95 +220,116 @@ int run_tests(const struct s2n_tls13_cert_verify_test *test_case, s2n_mode verif
-        EXPECT_SUCCESS(s2n_stuffer_write(&certificate_in, &b));
-
-        EXPECT_SUCCESS(s2n_stuffer_certificate_from_pem(&certificate_in, &certificate_out));
-
-        uint32_t available_size = s2n_stuffer_data_available(&certificate_out);
-        EXPECT_SUCCESS(s2n_blob_init(&b, s2n_stuffer_raw_read(&certificate_out, available_size), available_size));
+        /* Extract public key from certificate and set it for verifying connection */
         if (verifying_conn->mode == S2N_CLIENT) {
             EXPECT_OK(s2n_asn1der_to_public_key_and_type(&verifying_conn->handshake_params.server_public_key, &pkey_type, &b));
+            EXPECT_SUCCESS(s2n_pkey_match(&verifying_conn->handshake_params.server_public_key, sending_conn->handshake_params.our_chain_and_key->private_key));
         } else {
             EXPECT_OK(s2n_asn1der_to_public_key_and_type(&verifying_conn->handshake_params.client_public_key, &pkey_type, &b));
+            EXPECT_SUCCESS(s2n_pkey_match(&verifying_conn->handshake_params.client_public_key, sending_conn->handshake_params.our_chain_and_key->private_key));
         }
 
         /* Hash initialization */
+        EXPECT_SUCCESS(s2n_hash_init(&sending_conn->handshake.hashes->sha256, S2N_HASH_SHA256));
+        EXPECT_SUCCESS(s2n_hash_update(&sending_conn->handshake.hashes->sha256, hello, strlen((char *) hello)));
         EXPECT_SUCCESS(s2n_hash_init(&verifying_conn->handshake.hashes->sha256, S2N_HASH_SHA256));
         EXPECT_SUCCESS(s2n_hash_update(&verifying_conn->handshake.hashes->sha256, hello, strlen((char *) hello)));
 
-        /* Send and receive with mismatched hash algs */
-        EXPECT_SUCCESS(s2n_tls13_cert_verify_send(verifying_conn));
+        /* Send cert verify */
+        EXPECT_SUCCESS(s2n_tls13_cert_verify_send(sending_conn));
+        EXPECT_SUCCESS(s2n_stuffer_copy(&sending_conn->handshake.io, &verifying_conn->handshake.io, s2n_stuffer_data_available(&sending_conn->handshake.io)));
+        uint32_t write_cursor_position = s2n_stuffer_data_available(&verifying_conn->handshake.io);
 
-        /* Reinitialize hash */
-        EXPECT_SUCCESS(s2n_hash_init(&verifying_conn->handshake.hashes->sha256, S2N_HASH_SHA256));
-        EXPECT_SUCCESS(s2n_hash_update(&verifying_conn->handshake.hashes->sha256, hello, strlen((char *) hello)));
+        /* Use a hash algorithm different from sender by prepending corresponding iana value */
+        EXPECT_SUCCESS(s2n_stuffer_rewrite(&verifying_conn->handshake.io));
+        if (sig_scheme.sig_alg == S2N_SIGNATURE_ECDSA) {
+            EXPECT_SUCCESS(s2n_stuffer_write_uint16(&verifying_conn->handshake.io, TLS_SIGNATURE_SCHEME_ECDSA_SHA384));
+        } else {
+            EXPECT_SUCCESS(s2n_stuffer_write_uint16(&verifying_conn->handshake.io, TLS_SIGNATURE_SCHEME_RSA_PSS_PSS_SHA384));
+        }
+        EXPECT_SUCCESS(s2n_stuffer_skip_write(&verifying_conn->handshake.io, write_cursor_position));
```

Comment:
It's a shame that RSA-PSS was added in TLS1.3 so doesn't follow the TLS1.2 formula of the first byte being the hash and the second byte being the sig alg. That would have made this test easier :( https://datatracker.ietf.org/doc/html/rfc5246#section-7.4.1.4.1

But this is incredibly fragile as-is. You need a sig scheme that matches for one field but not for the other. You should probably add a field to the s2n_tls13_cert_verify_test struct so that we can specify the "sig scheme with different hash" when we specify the test sig scheme.